### PR TITLE
Adds Python bindings for the Light convenience class

### PR DESCRIPTION
**Pull request by Voldivh**
# 🎉 New feature

## Summary
This PR adds bindings for the `Light` convenience class.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: python/src/gz/sim/Light.cc [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  py::class_<gz::sim::Light>(module, "Light")
+  .def(py::init<gz::sim::Entity>())
+  .def("entity", &gz::sim::Light::Entity,
+      "Get the entity which this light is related to.")
+  .def("reset_entity", &gz::sim::Light::ResetEntity,
+      "Reset Entity to a new one.")
+  .def("valid", &gz::sim::Light::Valid,
+      py::arg("ecm"),
+      "Check whether this light correctly refers to an entity that"
+      "has a components::Light.")
```
Comment (ahcorde) [inline python/src/gz/sim/Light.cc:44]:
```suggestion
      "Check whether this light correctly refers to an entity that "
      "has a components::Light.")
```

File: python/src/gz/sim/Light.cc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  .def("name", &gz::sim::Light::Name,
+      py::arg("ecm"),
+      "Get the light's unscoped name.")
+  .def("pose", &gz::sim::Light::Pose,
+      py::arg("ecm"),
+      "Get the pose of the light."
+      "The pose is given w.r.t the light's parent. which can be a world or"
+      "a link.")
```
Comment (ahcorde) [inline python/src/gz/sim/Light.cc:52]:
```suggestion
      "The pose is given w.r.t the light's parent. which can be a world or "
      "a link.")
```

File: python/src/gz/sim/Light.cc [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  .def("intensity", &gz::sim::Light::Intensity,
+      py::arg("ecm"),
+      "Get the light intensity.")
+  .def("direction", &gz::sim::Light::Direction,
+      py::arg("ecm"),
+      "Get the light direction.")
+  .def("attenuation_range", &gz::sim::Light::AttenuationRange,
+      py::arg("ecm"),
+      "Get the light attenuation range."
+      "Light attenuation is not applicable to directional lights.")
```
Comment (ahcorde) [inline python/src/gz/sim/Light.cc:74]:
```suggestion
      "Get the light attenuation range. "
      "Light attenuation is not applicable to directional lights.")
```

File: python/src/gz/sim/Light.cc [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  .def("spot_inner_angle", &gz::sim::Light::SpotInnerAngle,
+      py::arg("ecm"),
+      "Get the inner angle of light. Applies to spot lights only.")
+  .def("spot_outer_angle", &gz::sim::Light::SpotOuterAngle,
+      py::arg("ecm"),
+      "Get the outer angle of light. Applies to spot lights only.")
+  .def("spot_fall_off", &gz::sim::Light::SpotFalloff,
+      py::arg("ecm"),
+      "Get the fall off value of light. Applies to spot lights only.")
```
Comment (ahcorde) [inline python/src/gz/sim/Light.cc:95]:
```suggestion
      "Get the light attenuation constant value. "
      "Light attenuation is not applicable to directional lights.")
  .def("attenuation_linear", &gz::sim::Light::AttenuationLinear,
      py::arg("ecm"),
      "Get the light attenuation linear value. "
      "Light attenuation is not applicable to directional lights.")
  .def("attenuation_quadratic", &gz::sim::Light::AttenuationQuadratic,
      py::arg("ecm"),
      "Get the light attenuation quadratic value. "
      "Light attenuation is not applicable to directional lights.")
  .def("spot_inner_angle", &gz::sim::Light::SpotInnerAngle,
      py::arg("ecm"),
      "Get the inner angle of light. Applies to spot lights only.")
  .def("spot_outer_angle", &gz::sim::Light::SpotOuterAngle,
      py::arg("ecm"),
      "Get the outer angle of light. Applies to spot lights only.")
  .def("spot_fall_off", &gz::sim::Light::SpotFalloff,
      py::arg("ecm"),
      "Get the fall off value of light. Applies to spot lights only.")
```

File: python/src/gz/sim/Light.hh [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
```
Comment (ahcorde) [inline python/src/gz/sim/Light.hh:2]:
```suggestion
 * Copyright (C) 2023 Open Source Robotics Foundation
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
tests?

File: python/src/gz/sim/Light.cc (hunk 3 above)
Comment (Voldivh) [reply python/src/gz/sim/Light.cc]: Done.

File: python/src/gz/sim/Light.cc (hunk 2 above)
Comment (Voldivh) [reply python/src/gz/sim/Light.cc]: Done.

**Comment by Voldivh** [comment]
@azeey The setters for this class doesn't seem to be working for some reason I'm not able to find. The test I added shouldn't be passing after I set the property to another value and they do. Also there are other values that doesn't change from the default value even after defining it on the sdf file such as the `spot_inner_angle`, `spot_outer_angle` and `spot_falloff`.

File: python/test/light_test.sdf [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" ?>
+<sdf version="1.6">
+  <world name="world_test">
+
+    <light type="point" name="light_test">
```
Comment (azeey) [inline python/test/light_test.sdf:5]: The `<direction>` and `<spot>` parameters do not apply when the light type is "point" (see http://sdformat.org/spec?ver=1.10&elem=light).

Comment (Voldivh) [reply python/test/light_test.sdf:5]: Oh I see, we have `point`, `spot` and `directional`. That makes a lot more sense. That should solve some of my problems, but, I think the setters issue remains the same.

Comment (Voldivh) [reply python/test/light_test.sdf]: I made the required changes in https://github.com/gazebosim/gz-sim/pull/2043/commits/4e72d74d66f78bfb3bfa0145198d1dea9033c0d9. However, the behavior is the same as I expected, solved some issues but not a single `set` method is working.

File: python/src/gz/sim/Light.hh [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <gz/sim/Light.hh>
+
+namespace gz
+{
+namespace sim
+{
+namespace python
+{
+/// Define a pybind11 wrapper for a gz::sim::World
```
Comment (ahcorde) [inline python/src/gz/sim/Light.hh:30]:
```suggestion
/// Define a pybind11 wrapper for a gz::sim::Light
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: python/src/gz/sim/Light.cc [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (azeey) [reply python/src/gz/sim/Light.cc:21]: @Voldivh can you address this?

File: python/test/light_TEST.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            # Pose Test
+            self.assertEqual(Pose3d(0, 2, 2, 0, 0, 0), light_point.pose(_ecm))
+            light_point.set_pose(_ecm, Pose3d(4, 2, 2, 0, 0, 0))
+            # Type Test
+            self.assertEqual('point', light_point.type(_ecm))
+            self.assertEqual('directional', light_directional.type(_ecm))
+            self.assertEqual('spot', light_spot.type(_ecm))
+            ## Diffuse Color Test
+            self.assertEqual(Color(1, 0, 0, 1), light_point.diffuse_color(_ecm))
+            light_point.set_diffuse_color(_ecm, Color(1, 1, 0, 1))
```
Comment (azeey) [inline python/test/light_TEST.py:69]: Is there supposed to be a test here affter seting the diffuse color to check that it worked? Same for specular, cast shadows, etc.

File: python/src/gz/sim/Light.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (Voldivh) [reply python/src/gz/sim/Light.cc:21]: Done.

File: python/src/gz/sim/Light.hh (hunk 7 above)
Comment (Voldivh) [reply python/src/gz/sim/Light.hh]: Thanks and done.

File: python/test/light_TEST.py (hunk 9 above)
Comment (Voldivh) [reply python/test/light_TEST.py:72]: I added them, however, they are not working for some reason, I pushed and example for you to take a look.

Comment (azeey) [reply python/test/light_TEST.py:72]: @iche033 It looks like the getter (`Light::DiffuseColor`) returns the value of the `sdf::Light` which was in the original SDF file, not the current state of the light in the scene. Is that the intended behavior?

Comment (azeey) [reply python/test/light_TEST.py:72]: I've created an issue https://github.com/gazebosim/gz-sim/issues/2095. For now, let's just skip the tests for the `set*` functions.

Comment (azeey) [reply python/test/light_TEST.py:72]: It turns out the getters do work, but a rendering system, such as the Sensor system (and a camera) is required. I've removed the checks that require this in https://github.com/gazebosim/gz-sim/pull/2043/commits/209e5562d55cc21186bf11c9d02e15dc216841a7

**Review by azeey — APPROVED** [review]

**Review by ahcorde — CHANGES_REQUESTED** [review]
There is still a pending comment, good to go when the `iostream` header is removed

File: python/src/gz/sim/Light.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (azeey) [reply python/src/gz/sim/Light.cc]: Might have been re-added when bringing back an older commit for testing. Fixed in https://github.com/gazebosim/gz-sim/pull/2043/commits/801bb087a9426259c83e573d6c7228fcf37f67be

**Review by ahcorde — APPROVED** [review]

**Pull request merged**